base/compiler/abstractlattice.jl

```diff
@@ -0,0 +1,127 @@
+abstract type AbstractLattice; end
+function widen end
+
+"""
+    struct JLTypeLattice
+
+A singleton type representing the lattice of Julia types, without any inference
+extensions.
+"""
+struct JLTypeLattice <: AbstractLattice; end
+widen(::JLTypeLattice) = error("Type lattice is the least-precise lattice available")
+
+"""
+    struct ConstsLattice
+
+A lattice extending `JLTypeLattice` and adjoining `Const` and `PartialTypeVar`.
+"""
+struct ConstsLattice <: AbstractLattice; end
+widen(::ConstsLattice) = JLTypeLattice()
+
+"""
+    struct PartialsLattice{L}
+
+A lattice extending lattice `L` and adjoining `PartialStruct` and `PartialOpaque`.
+"""
+struct PartialsLattice{L <: AbstractLattice} <: AbstractLattice
+    parent::L
+end
+widen(L::PartialsLattice) = L.parent
+
+"""
+    struct PartialsLattice{L}
+
+A lattice extending lattice `L` and adjoining `Conditional`.
+"""
+struct ConditionalsLattice{L <: AbstractLattice} <: AbstractLattice
+    parent::L
+end
+widen(L::ConditionalsLattice) = L.parent
+
+"""
+    struct OptimizerLattice
+
+The lattice used by the optimizer. Equivalent to
+`ConditionalsLattice{PartialsLattice{ConstsLattice}}`, but wrapped in a struct.
+"""
+struct OptimizerLattice <: AbstractLattice; end
+widen(L::OptimizerLattice) = ConditionalsLattice(PartialsLattice(ConstsLattice()))
+
+"""
+    struct InferenceLattice
+
+The full lattice used for abstract interpration during inference. This takes the
+`OptimizerLattice` and adjoints `LimitedAccuracy` and `MaybeUndef`.
```

Comment:
`MaybeUndef` only appears in optimization, so I think it should be handled by `OptimizerLattice`?

Reply:
:+1: So I guess `LimitedAccuracy` doesn't show up in the optimizer and `MaybeUndef` doesn't show up in the inference lattice, so I'll make them separate but both wrap the conditionals lattice.